sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1041,10 +1054,11 @@ public boolean start() throws IOException {
           consumer.seek(p.topicPartition, p.nextOffset);
         } else {
           // nextOffset is unininitialized here, meaning start reading from latest record as of now
-          // ('latest' is the default, and is configurable). Remember the current position without
-          // waiting until the first record read. This ensures checkpoint is accurate even if the
-          // reader is closed before reading any records.
-          p.nextOffset = consumer.position(p.topicPartition);
+          // ('latest' is the default, and is configurable) or look up the offset by startTime.
+          // Remember the current position without waiting until the first record read. This
+          // ensures checkpoint is accurate even if the reader is closed before reading any records.
+          p.nextOffset =
+              consumerSpEL.offsetForTime(consumer, p.topicPartition, spec.getStartReadTime());
```

Comment:
The default should not be to start at ReadTime(), unless it is explicitly set. You are kind of doing that in the implementation of offsetForTime(), but that does not seem like right contract. `offsetForTime()` should fail if the consumer could fetch offset of the time.  `consumer.position(p.topicPartition)` honors Kafka configuration set by users, we should handle the distinction explicitly here.